#159987728 Users can make reports on articles

emekafredy opened the pull request:
# What does this PR do?
Create endpoints for users to report articles and for admin to work on these reports

# Description of Task to be completed?
### Create ReportsOnArticle model and its association with users and articles
### Write unit tests for feature
### Create endpoints:
- Report article ``` POST api/articles/<article_slug>/report ```
- View all reports ``` GET api/admin/articles/reports ```
- View a single report ``` GET api/admin/articles/reports/<reportId>```
- View all reports for a single article ``` GET api/admin/reports/articles/<article_slug> ```
- Block article ``` PUT api/admin/articles/:article_slug/delist ```
- View blocked articles ``` GET api/admin/articles/delisted ```
- View a blocked article ``` GET api/admin/articles/:article_slug/delisted ```

- Unblock article ``` PUT api/admin/articles/:article_slug/relist ```
### Update swagger documentation for the above endpoints

# How should this be manually tested?
- Clone the repo
- cd into the 'Thor-ah'
- Run 'npm install'
- Use the .env.sample format to create a .env file
- Run 'npm run migrate'
- Run 'npm run start:dev'


# What are the relevant pivotal tracker stories?
[159987728](https://www.pivotaltracker.com/story/show/159987728)

emekafredy committed:
feat(reportArticles): endpoint to make reports on articles

- Create model for reports and associate to users and articles.
- Write tests for the model and features.
- Define endpoints for:
	- Users to report an article
	- Admin to see reports made on articles
	- Admin to block and unblock articles
	- Admin to view all blocked articles
- Update swagger documentation for comments and article reports

[Delivers #159987728]

CEOehis commented on server/controllers/adminHandleReports.js:
```diff
@@ -0,0 +1,224 @@
+      }]
+    })
+      .then(report => res.status(200).json({
+        report
```
You can add a `status` property with value `success` in your response. Same with line `99-102`

emekafredy replied:
Okay. Thanks for noticing this omission

CEOehis reviewed (changes requested):
You can add a `status` property with value `success` in your response in lines `67-69`. Same thing with line `99-102`

emmaadesile commented on server/controllers/adminHandleReports.js:
```diff
@@ -0,0 +1,224 @@
+      }]
+    })
+      .then(reports => res.status(200).json({
+        reports
```
You could modify the response as
```
res.status(200).json({
  status: 'success',
  report,
})
```

emmaadesile commented on server/controllers/adminHandleReports.js:
```diff
@@ -0,0 +1,224 @@
+      }]
+    })
+      .then(report => res.status(200).json({
+        report
```
Same as line `36, 68`

emmaadesile commented on server/controllers/article.js:
```diff
@@ -346,6 +361,69 @@ class ArticleController {
+              return res.status(400).json({
+                status: 'error',
+                error: {
+                  message: 'We already have your report on this article'
```
A user should be able to report an article more than once. Line `398-410`

emekafredy replied:
Okay. I guess that makes more sense

d-beloved commented on server/controllers/article.js:
```diff
@@ -346,6 +361,69 @@ class ArticleController {
+        if (!article) {
+          return res.status(404).json({
+            status: 'error',
+            errors: {
```
please change this to `error`

emekafredy replied:
Okay. Will do that immediately

emekafredy committed:
fix(reportArticles): implement requested changes on PR review

- Users should be allowed to make more than one report on a single article
- Add the status property for each returned promise.
- For uniformity, all error properties should read error, not ‘errors’.

[Delivers #159987728]

emmaadesile reviewed (approved):
LGTM

CEOehis reviewed (approved):
LGTM

aknwosu commented on server/controllers/adminHandleReports.js:
```diff
@@ -0,0 +1,225 @@
+   * @returns {next} next calls next
+   * @memberof HandleReports
+   */
+  static getAllReports(req, res, next) {
```
static getReportedArticles(req, res, next) {

aknwosu commented on server/controllers/adminHandleReports.js:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  /**
+   * @description get a single report made by a user on an article
```
* @description get a single reported article

aknwosu commented on server/controllers/adminHandleReports.js:
```diff
@@ -0,0 +1,225 @@
+   * @returns {next} next calls next
+   * @memberof HandleReports
+   */
+  static getSingleReport(req, res, next) {
```
getAReportedArticle

aknwosu commented on server/controllers/adminHandleReports.js:
```diff
@@ -0,0 +1,225 @@
+   * @returns {next} next calls next
+   * @memberof HandleReports
+   */
+  static getReportsForSingleArticle(req, res, next) {
```
getReportsForAnArticle

emekafredy committed:
fix(reportArticles): Change method names to convey more meaning

aknwosu committed:
Merge branch 'develop' into ft-reports-on-articles-endpoint-159987728
